unified-scheduler-logic/src/lib.rs

```diff
@@ -1,20 +1,1299 @@
+        if matches!(
+            self.blocked_usages_from_tasks.front(),
+            Some((RequestedUsage::Readonly, _))
+        ) {
+            self.pop_unblocked_usage_from_task()
+        } else {
+            None
+        }
+    }
+
+    fn has_no_blocked_usage(&self) -> bool {
+        self.blocked_usages_from_tasks.is_empty()
+    }
+}
+
+const_assert_eq!(mem::size_of::<TokenCell<UsageQueueInner>>(), 40);
+
+/// Scheduler's internal data for each address ([`Pubkey`](`solana_sdk::pubkey::Pubkey`)). Very
+/// opaque wrapper type; no methods just with [`::clone()`](Clone::clone) and
+/// [`::default()`](Default::default).
+#[derive(Debug, Clone, Default)]
+pub struct UsageQueue(Arc<TokenCell<UsageQueueInner>>);
+const_assert_eq!(mem::size_of::<UsageQueue>(), 8);
+
+/// A high-level `struct`, managing the overall scheduling of [tasks](Task), to be used by
+/// `solana-unified-scheduler-pool`.
+pub struct SchedulingStateMachine {
+    unblocked_task_queue: VecDeque<Task>,
+    active_task_count: ShortCounter,
+    handled_task_count: ShortCounter,
+    unblocked_task_count: ShortCounter,
+    total_task_count: ShortCounter,
+    count_token: BlockedUsageCountToken,
+    usage_queue_token: UsageQueueToken,
+}
+const_assert_eq!(mem::size_of::<SchedulingStateMachine>(), 48);
+
+impl SchedulingStateMachine {
+    pub fn has_no_active_task(&self) -> bool {
+        self.active_task_count.is_zero()
+    }
+
+    pub fn has_unblocked_task(&self) -> bool {
+        !self.unblocked_task_queue.is_empty()
+    }
+
+    pub fn unblocked_task_queue_count(&self) -> usize {
+        self.unblocked_task_queue.len()
+    }
+
+    pub fn active_task_count(&self) -> u32 {
+        self.active_task_count.current()
+    }
+
+    pub fn handled_task_count(&self) -> u32 {
+        self.handled_task_count.current()
+    }
+
+    pub fn unblocked_task_count(&self) -> u32 {
+        self.unblocked_task_count.current()
+    }
+
+    pub fn total_task_count(&self) -> u32 {
+        self.total_task_count.current()
+    }
+
+    /// Schedules given `task`, returning it if successful.
+    ///
+    /// Returns `Some(task)` if it's immediately scheduled. Otherwise, returns `None`,
+    /// indicating the task is blocked currently.
+    ///
+    /// Note that this function's type signature is intentionally redundant to take the ownership
+    /// of given task _conditionally_ for future optimization.
+    #[must_use]
+    pub fn schedule_task(&mut self, task: Task) -> Option<Task> {
+        self.total_task_count.increment_self();
+        self.active_task_count.increment_self();
+        self.try_lock_for_task(task)
+    }
+
+    #[must_use]
+    pub fn schedule_unblocked_task(&mut self) -> Option<Task> {
+        self.unblocked_task_queue.pop_front().inspect(|_| {
+            self.unblocked_task_count.increment_self();
+        })
+    }
+
+    pub fn deschedule_task(&mut self, task: &Task) {
```

Comment:
good question: 250edde6dde10899ffb7c05d1b143fca4b0f1705

> If not, the caller can re-schedule the same task, which seems wrong API wise

Even if this fn consumes the task, this would be possible because `Task` is `Clone`... That said, I'm planning to make it `!Clone`.

Reply:
From the referenced commit 

>     /// Note that this function intentionally doesn't take ownership of the task to avoid dropping
>    /// tasks inside `SchedulingStateMachine` to provide an offloading-based optimization
 >   /// opportunity for callers.

Could you say more about what optimization you have in mind? Documenting invariants is good, but type system > docs, so if we can, we should enforce this at the type level IMO.

(And agreed that tasks shouldn't be Clone)

Reply:
> Could you say more about what optimization you have in mind?

well, i thought i was explicit. avoiding dropping and offloading that to other background thread is an optimization. as a general context, the scheduler thread itself is the weakest point (bottleneck) of my unified scheduler thing. its performance is very susceptible to any overhead...

Reply:
btw, the optimization will follow soon because I'm now in to mood of missing 1.19 (or 2.0?) branching.

> (And agreed that tasks shouldn't be Clone)

btw, this will be fixed together.

Reply:
> well, i thought i was explicit. avoiding dropping and offloading that to other background thread is an optimization. 

It's not explicit because from the existing code it's not obvious what the optimization would even be. Currently the thread calling deschedule_task() is in fact dropping the Task (presumably - because to further complicate things, Task is an Arc, so who knows who's holding refs). This btw is why personally I'm not a fan of landing big changes piecemeal - it's easier to review code in its proper context than in isolation and then "things will happen in the future".

Anyway, whatever is happening to the task, surely this is something that can be expressed through the type system? Eg if after deschedule_task(task) the task can't be scheduled anymore, deschedule_task could take by value and return a marker Completed(task) type which makes it impossible to access the inner value. Then Completed(task) can still be dropped in another thread.

Reply:
> It's not explicit because from the existing code it's not obvious what the optimization would even be. Currently the thread calling deschedule_task() is in fact dropping the Task (presumably - because to further complicate things, Task is an Arc, so who knows who's holding refs). This btw is why personally I'm not a fan of landing big changes piecemeal - it's easier to review code in its proper context than in isolation and then "things will happen in the future".

yeah. sorry for extra trouble... as you noticed, it's currently dropping in-thread indeed. and that's because of piecemeal approach.. i know these back-and-fort aren't fun. however, i still like this approach as personal preference.

> Anyway, whatever is happening to the task, surely this is something that can be expressed through the type system? ...

yep, i agree. thanks for suggestion. I'll do once after https://github.com/anza-xyz/agave/pull/129#issuecomment-2016423148 is settled on not trashing this whole involved approch in this pr...